### PR TITLE
+ akka-persistence: add internal message protocol traits, resolve #16612

### LINKED ISSUE #16612: akka-persistence: JournalProtocol/SnapshotProtocol should extend 3 marker traits
Currently [JournalProtocol](https://github.com/akka/akka/blob/master/akka-persistence/src/main/scala/akka/persistence/JournalProtocol.scala) and [SnapshotProtocol] (https://github.com/akka/akka/blob/master/akka-persistence/src/main/scala/akka/persistence/Snapshot.scala#L115) do not extend any  marker traits.

This is problematic for persistence plugin development, when custom Journal and Snapshot actors
need process other in-band commands, such as cluster replication.

[For example, here](https://github.com/carrot-garden/akka-persistence-chronicle/blob/master/src/main/scala/akka/persistence/chronicle/Journal.scala#L273) wildcard case will potentially forward wrong commands down to the persistence supertype.

Solution is to provide 3 marker traits and have relevant messages extend them:
```
sealed trait PersistenceCommand
sealed trait PersistenceJournalCommand
sealed trait PersistenceSnapshotCommand
```

### PR DESCRIPTION
**Pull request by Andrei-Pozolotin**

### CODE REVIEW

**Comment by Andrei-Pozolotin** [comment]
@patriknw please review

**Comment by patriknw** [comment]
OK TO TEST

File: akka-persistence/src/main/scala/akka/persistence/package.scala [hunk 1]
```diff
@@ -12,3 +12,17 @@ package object persistence {
...
+   * INTERNAL API.
+   *
+   * Internal persistence extension messages extend this trait.
+   *
+   * Helps persistence plugin developers to differentiate
+   * internal persistence extension messages from their custom plugin messages.
+   */
+  private[persistence] trait PersistenceMessage
+
+}
```
Comment (patriknw) [inline akka-persistence/src/main/scala/akka/persistence/package.scala]: I would prefer that you define this trait as a top level trait in `JournalProtocol.scala`

**Comment by patriknw** [comment]
Refs #16612

**Comment by patriknw** [comment]
LGTM, aside from the comment

**Comment by patriknw** [comment]
thanks

**Comment by patriknw** [comment]
@ktoso please take a look at this one

File: akka-persistence/src/main/scala/akka/persistence/package.scala (hunk 1 above)
Comment (ktoso) [inline akka-persistence/src/main/scala/akka/persistence/package.scala]: +1, it's surprising here

File: akka-persistence/src/main/scala/akka/persistence/JournalProtocol.scala [hunk 2]
```diff
@@ -14,10 +14,19 @@ import akka.actor._
  * Messages exchanged between persistent actors, views and a journal.
  */
 private[persistence] object JournalProtocol {
+
+  /** Marker trait shared by internal journal messages. */
+  sealed trait Message extends PersistenceMessage
```
Comment (ktoso) [inline akka-persistence/src/main/scala/akka/persistence/JournalProtocol.scala]: +1 for these short classnames since in `JournalProtocol` already

**Comment by ktoso** [comment]
LGTM, but would like the `PersistenceMessage` moved to top-level or into `object Persistence`.

File: akka-persistence/src/main/scala/akka/persistence/package.scala (hunk 1 above)
Comment (Andrei-Pozolotin) [inline akka-persistence/src/main/scala/akka/persistence/package.scala]: relocated to separate unit

File: akka-persistence/src/main/scala/akka/persistence/JournalProtocol.scala (hunk 2 above)
Comment (Andrei-Pozolotin) [inline akka-persistence/src/main/scala/akka/persistence/JournalProtocol.scala]: ok

**Comment by Andrei-Pozolotin** [comment]
re: "moved to top-level " - relocated to separate unit

**Comment by patriknw** [comment]
Thanks. Please squash to one commit, and we will merge.

**Comment by patriknw** [comment]
also, write the commit description according to the conventions in https://github.com/akka/akka/blob/master/CONTRIBUTING.md

In this case something like:

    +per #16612 add internal message protocol traits

**Comment by Andrei-Pozolotin** [comment]
squashed

**Comment by Andrei-Pozolotin** [comment]
thank you

**Pull request merged**